### ISSUE TITLE
Bug: Setting content-type on reply object does not work correctly and/or hangs server

### ISSUE THREAD

**Issue by petef19** [fastify/fastify#3283]
## 💬  Bug report

It seems that `fastify` tries to force guess the `content-type` of the response, and does ONLY allow to override it on the `reply` object if it matches what it has guessed/determined, but even that does not work correctly all the time.

example 1
```
reply
    .code(200)
    .headers({'content-type': 'text/plain; charset=utf-32'})
    .send(JSON.stringify({bar: 'foo', baz: 'foobar'}));
````
this works b/c `.send` is sending a text string, hence `fastify` will take the override in `.headers()`. Btw, the override also works in this example w/o adding the `charset` (contrary to example 3 below).

example 2
```
reply
    .code(200)
    .headers({'content-type': 'text/plain; charset=utf-8'})
    .send({bar: 'foo', baz: 'foobar'});
````
this does NOT work - `.send` is sending an object, hence `fastify` determines `application/json; charset=utf-8` and will simply **ignore the override** in `.headers()`... more importantly: **this now hangs the server** forever, the response will never come through

example 3
```
reply
    .code(200)
    .headers({'content-type': 'application/json'})
    .send({bar: 'foo', baz: 'foobar'});
```
looking at the code you'd think this would work, we're sending an object and `fastify` will guess `application/json` and hence use the override in `.headers()`. It doesn't. It sends `application/json; charset=utf-8`, which does NOT match what was specified in `.headers()`.

Summary:
example 1: it uses the override
example 2: it does NOT use the override and hangs
example 3: it does NOT use the override although the logic is very similar to example 1

## Your Environment

- *node version*: 14.6.0
- *fastify version*: 3.20.2
- *os*: Win 10 x64

**Comment by climba03003** [comment]
We do have 2 test-case against the invalid payload, however it is checked inside the `handler` which do not ensure the client always get the error response.

https://github.com/fastify/fastify/blob/a04f4c952731c57e79c6ca3794a62959a43731ed/test/reply-error.test.js#L398-L445

**Comment by petef19** [comment]
> Example 2:
> You need to ensure the payload is one of the following type which stated in the document clearly.
> https://github.com/fastify/fastify/blob/main/docs/Reply.md#type-of-the-final-payload

what is confusing about this is that `fastify` usually converts objects automatically to `json` - in example 2, if I omit setting the header, it automatically stringifies the object and then sets `application/json` - but I guess it doesn't do it when I set 'plain/text';

> Example 3:
> It is intended behavior to always set charset when we see application/json only content-type.

ok, hopefully that is in the docs somewhere. 
But, what is confusing is, that with `plain/text` content-type  (set by user) `fastify` does NOT automatically add the `charset`, although when it auto detects/guesses `plain/text` (and user did not explicitly set a content type) it auto adds the `charset` to `plain/text` resulting in to `plain/text; charset=utf-8`....

Thanks !

**Comment by iamdevelopergirl** [comment]
I understand the problem and I want to work on this issue. Can I take it?

**Comment by mcollina** [comment]
> I understand the problem and I want to work on this issue. Can I take it? 

go for it!

**Comment by iamdevelopergirl** [comment]
Thanks.

**Comment by jsumners** [comment]
I think this can be closed as resolved by #3285.
